server/src/main/java/io/spine/server/entity/VersionIncrement.java

```diff
@@ -22,41 +22,92 @@
 
-import static io.spine.core.Versions.checkIsIncrement;
+import static com.google.common.base.Preconditions.checkNotNull;
 
 /**
- * A setter of a new version for the entity in transaction.
+ * A strategy for incrementing a version of an entity being modified in a transaction.
  */
 @Internal
 public abstract class VersionIncrement {
 
-    private final Transaction transaction;
+    /**
+     * Creates a new version for the entity in transaction.
+     *
+     * @return the incremented {@code Version} of the entity
+     */
+    protected abstract Version nextVersion();
 
-    VersionIncrement(Transaction transaction) {
-        this.transaction = transaction;
+    /**
+     * Creates a version increment which sets the new version from the given event.
+     *
+     * <p>Such increment strategy is applied to the {@link Entity} types which represent
+     * a sequence of events.
+     *
+     * <p>One example of such entity is {@link io.spine.server.aggregate.Aggregate Aggregate}.
+     * As a sequence of events, an {@code Aggregate} has no own versioning system, thus inherits
+     * the versions of the {@linkplain io.spine.server.aggregate.Apply applied} events.
+     * In other words, the current version of an {@code Aggregate} is the version of the last
+     * applied event.
+     */
+    public static VersionIncrement fromEvent(EventEnvelope e) {
+        return new IncrementFromEvent(e);
     }
 
     /**
-     * Advances the version of the entity in transaction.
+     * Creates a version increment which always advances the given entity version by 1.
+     *
+     * <p>Such increment strategy is applied to the {@link Entity} types which cannot use the event
+     * versions, such as {@link io.spine.server.projection.Projection Projection}s.
+     *
+     * <p>A {@code Projection} represents an arbitrary cast of data in a specific moment in
+     * time. The events applied to a {@code Projection} are produced by different {@code Entities}
+     * and have no common versioning. Thus, a {@code Projection} has its own versioning system.
+     * Each event <i>increments</i> the {@code Projection} version by one.
      */
-    void apply() {
-        Version nextVersion = nextVersion();
-        checkIsIncrement(transaction.version(), nextVersion);
-        transaction.setVersion(nextVersion);
+    public static VersionIncrement sequentially(Transaction tx) {
+        return new AutoIncrement(tx);
     }
 
     /**
-     * Returns the transaction on which the {@code VersionIncrement} operates.
+     * Increments version from a passed event.
      */
-    Transaction transaction() {
-        return transaction;
+    private static final class IncrementFromEvent extends VersionIncrement {
+
+        private final EventEnvelope event;
+
+        private IncrementFromEvent(EventEnvelope event) {
+            super();
+            this.event = event;
+        }
+
+        @Override
+        public Version nextVersion() {
```

Comment:
Please check if this method really should be `public` in a `private static final` class. The parent method is `protected`.